Compiler/src/verifytrim.jl

```diff
@@ -49,76 +49,223 @@ const runtime_functions = Symbol[
 
 ## code for pretty printing ##
 
-# wrap a statement in a typeassert for printing clarity, unless that info seems already obvious
-function mapssavaluetypes(codeinfo::CodeInfo, sptypes::Vector{VarState}, stmt)
-    @nospecialize stmt
-    newstmt = mapssavalues(codeinfo, sptypes, stmt)
-    typ = widenconst(argextype(stmt, codeinfo, sptypes))
-    if newstmt isa Expr
-        if newstmt.head ∈ (:quote, :inert)
-            return newstmt
+# Check if a union is small with concrete types (same logic as code_warntype's is_expected_union)
+function is_expected_union(@nospecialize(u::Union))
+    Base.unionlen(u) < 4 || return false
+    for x in Base.uniontypes(u)
+        if !isdispatchelem(x) || x == Core.Box
+            return false
         end
-    elseif newstmt isa GlobalRef && isdispatchelem(typ)
-        return newstmt
-    elseif newstmt isa Union{Int, UInt8, UInt16, UInt32, UInt64, Float16, Float32, Float64, String, QuoteNode}
-        return newstmt
-    elseif newstmt isa Callable
-        return newstmt
     end
-    return Expr(:(::), newstmt, typ)
+    return true
+end
+
+# Check if a type is "stable" - uses same classification as code_warntype
+function is_type_stable(@nospecialize(typ))
+    typ isa Core.Const && return true
+    typ === Union{} && return false
+    typ == Core.Box && return false
+    typ isa Union && return is_expected_union(typ)
+    return isdispatchelem(typ)
+end
+
+# Color printing for types - same colors as code_warntype:
+# - red bold: unstable/abstract types
+# - yellow: expected unions (small unions of concrete types)
+# - light_black: stable concrete types
+function print_type_colored(io::IO, @nospecialize(typ))
+    color = get(io, :color, false)::Bool
+    if !color
+        print(io, typ)
+    elseif typ === Union{}
+        printstyled(io, typ; color=:red)
+    elseif typ == Core.Box
+        printstyled(io, typ; color=:red, bold=true)
+    elseif typ isa Union && is_expected_union(typ)
+        printstyled(io, typ; color=:yellow)
+    elseif isdispatchelem(typ)
+        printstyled(io, typ; color=:light_black)
+    else
+        printstyled(io, typ; color=:red, bold=true)
+    end
+end
+
+# Print with light_black if stable, normal otherwise
+function print_styled(io::IO, @nospecialize(x), stable::Bool)
+    stable ? printstyled(io, x; color=:light_black) : print(io, x)
+end
+
+# Check if we should elide the type annotation for a statement
+function should_elide_type(newstmt, @nospecialize(typ))
+    newstmt isa Expr && newstmt.head ∈ (:quote, :inert) && return true
+    newstmt isa GlobalRef && isdispatchelem(typ) && return true
+    newstmt isa Union{Int, UInt8, UInt16, UInt32, UInt64, Float16, Float32, Float64, String, QuoteNode} && return true
```

Comment:
This is a really random set of types - Any justification for it?

I thought this might be either:
  - Base types that already print their type (e.g. `Float16`) or are supported as literals
  - Objects that are self-quoting

but it doesn't seem to match either

Reply:
These were already in the code tbf. The diff is messy

Reply:
I did not realize that! In that case, I won't bother you about it 🙂